Source/Provers/SMTLib/SMTLibProcess.cs

```diff
@@ -14,16 +14,18 @@ public class SMTLibProcess : SMTLibSolver
     readonly Process prover;
     readonly Inspector inspector;
     readonly SMTLibProverOptions options;
+    private readonly CancellationToken cancellationToken;
     readonly Queue<string> proverOutput = new();
     readonly Queue<string> proverErrors = new();
     readonly TextWriter toProver;
     readonly int smtProcessId;
     static int smtProcessIdSeq = 0;
     ConsoleCancelEventHandler cancelEvent;
 
-    public SMTLibProcess(SMTLibOptions libOptions, SMTLibProverOptions options)
+    public SMTLibProcess(SMTLibOptions libOptions, SMTLibProverOptions options, CancellationToken cancellationToken)
```

Comment:
I'm surprised to see a cancellationToken in a constructor. I'd expect a cancellationToken to be passed to a method that return a `Task`, or otherwise is a blocking method, where the semantics of the token would be "if you cancel it, this asynchronous method will return immediately". For a constructor that already returns immediately, it's less clear what the meaning of the token is.

Have you considered adding the token as an argument to `GetProverResponse`? That way you can also cancel even if the solver process is not stuck, but for some other reason you don't want to wait for the current response to be returned, and then you might re-use that solver later on and pass in a new cancellation token when calling `GetProverResponse`.

Similar, could you pass the CancellationToken down through from ExecutionEngine down through VerifyImplementation towards `SplitAndVerifyWorker.StartCheck`
split.BeginCheck, instead of through constructors?

Reply:
I was not aware of this habit to put CancellationTokens only in methods that return Task.

I did consider to put the token as an argument to `GetProverResponse()` but I did not chose that option, because I do want the process to be killed if the task is canceled, not reused. If the `CancellationToken.ThrowIfRequested()` is called, it means that we were in the middle of a computation that took longer to execute than the patience of the user (or, in the case of the Dafny IDE plugin, the user changed the file by 1 character and verification is on every change, which would completely invalidate the verification results).
Since we have no way to know how longer the process needs time to execute, I don't think any library should rely on a process that just got canceled. In the case of Dafny, reusing a canceled process would simply be like starting a new process, except that you have to wait an extra time between 0 and infinite until the previous verification finishes.

To conclude, if the CancellationToken is not in the calling method, it has to be a field, hence why it appears in the constructor, but you already figured that out. Let me know if you have more insights to share that might shake the reasoning above.

Reply:
I'm fine with keeping the current behavior, but I would like to only pass CancellationTokens as part of asynchronous methods. I think you can do both.

Reply:
> I would like to only pass CancellationTokens as part of asynchronous or blocking methods to make the meaning of the tokens clear.
A constructor is a blocking method; can you please let me know precisely what you mean?

Reply:
With blocking method I meant a method that does some time consuming operation such as IO and only returns after it's done with that operation. An example is `SMTLibProcess.ReadProver` since it calls `Monitor.Wait(this, 100)`. I think the idiomatic signature of this method would be `Task<string> ReadProver(CancellationToken cancellationToken)` which I think would also allow it to return immediately when `OutputDataReceived` fires without having to use `Monitor.Wait`. Changing the return type to `Task<string`> is a hassle but we can already add the CancellationToken argument in this PR.

Reply:
I agree it's idiomatic, but look at the changes it triggers. in the case of the prover (the last commit I just pushed)
https://github.com/boogie-org/boogie/pull/491/commits/528025538f0258be8cff7bbf6116b9ad15df51af
I would then add that, when the same unchanged argument is given over and over to several methods of the same class, it is also idiomatic to put it in a constructor so that the signatures of methods stays as simple as possible.

Reply:
The changes in that commit look good to me. I think with a different design a .NET `Task` might have been equivalent to `Func<CancellationToken, WorkTask>`, similar to a [State monad](https://wiki.haskell.org/State_Monad), and then you wouldn't need to propagate cancellationTokens, but with the current design, lots of propagation of cancellationTokens is the way it is.

But em.. thinking more about this. I'm confused you even need to do `cancellationToken.ThrowIfCancellationRequested();` there. Shouldn't the `!prover.HasExited` in `while (proverOutput.Count == 0 && proverErrors.Count == 0 && !prover.HasExited)` already allow it to return when the process gets killed, which should happen when `ExecutionEngine.CancelRequest` is called. What's going on?

Reply:
That's a brilliant finding. HIstorically, I first threw the error there so that i would stop the loop, and later I added `cancellationToken.Register(KillProcess);` in the constructor of `SMTLibProcessTheoremProver` to fix another error.
But it looks like when I comment `cancellationToken.ThrowIfCancellationRequested()` because `cancellationToken.Register(KillProcess);` does the work.

I pushed a commit that removes the cancellationTokens inside the process alltogether

Reply:
But, why do we need the changes in this PR at all, since `SMTLibProcess.Close` should already be called as a result of `ExecutionEngine.CancelRequest` being called?

Also, I don't think we should merge this PR without it having a test.

Reply:
After a lot of further investigation, here is what happens:

`ExecutionEngine.CancelRequest` calls `CleanupCheckers(requestId)`
`CleanupCheckers(requestId)` calls `checkerPool?.Dispose();` **only if there is no other pending request**
`checkerPool.Dispose();` calls `checker.Close();` **only for checkers that are available (idle)**

Here we see why it cannot work without `cancellationToken.Register(KillProcess);`. First, if there are other pending requests, the `checkerPool` is not disposed. Second, disposing the `checkerPool` only calls `Close` on the checkers that were available. However, one of the two checkers that were created in my test did not finish, so it was never registered to be available and does not close.

This PR has a test but it is in Dafny:
https://github.com/dafny-lang/dafny/pull/1771
I can convert it to Boogie, but can you please let me know how you would proceed for that? There is no asynchronous testing anywhere, so I'm not sure where to start.

Reply:
The use of `Task.WaitASync(CancellationToken)` is very interesting, one, thanks for your suggestion, and let me have a deeper look at it.
I also tested your branch, and right now the Dafny Language Server test succeeds, but as expected the process that started the test does not finish because Z3 is not killed. See the attached screenshot.
![image](https://user-images.githubusercontent.com/3601079/153265664-93c0a44a-97cd-4179-9f26-22ba3293cbe1.png)

I suggest we do both what you are doing and what I am doing, so that we can end even faster for the client (with your change, no need to wait the extra 2000ms if the prover is unresponsive)

>You might say that starting Z3 processes is cheap so we don't need to recycle them, but then why do we have a CheckerPool class in the first place?

I do not have a definitive answer, but I think this is an excellent question to ask now.
Here are some insights I gathered while looking at CheckerPool.cs and Checker.cs just now:
* CheckerPool ensures that there are at most <number of VCSCores> checkers in the pool. This is very clever, because we don't want more than that processes running at the same time.
  However, CheckerPool **does not keep a pointer on checkers**, only an abstract representation `int notCreatedCheckers ` of how many checkers it can still create.
* When something needs a checker, `CheckerPool.FindCheckerFor` will return a promise for a checker (a Task<Checker>). This promise is fullfilled by either
  * Detecting if an existing checker is idle, meaning there is a checker in `availableCheckers`, meaning something called the checker's `GoBackToIdle()` method (in two places, one being the end of `SplitAndVerifyWorker.DoWork` in the `finally` you refactored when it calls `split.ReleaseChecker();`) which normally ensures that the z3 process answered and can now listen to standard input
  * If the number of maximum checkers has not been reached, creating a new checker and returning it immediately as a `Task.FromResult(checker)`
  * Else, it **assumes that a checker will eventually get idle**, and enqueue the request for a checker in `checkerWaiters`, and assigns it the continuation to `PrepareChecker` on the obtained checker when it will be available, and returning the checker itself. This request is fullfilled in `Checker.GoBackToIdle`, which in turns call `CheckerPool.AddChecker`, which, if there is a waiter in `checkerWaiters`, immediately provide the checker, and if nothing is waiting for a checker, it will enqueue the idle checker in `availableCheckers`

So here is a updated comment now with this updated information:
When a cancellation is requested, there is no way to query the process to say "Hey, please cancel this request" nicely. Moreover, it's even possible that the process will never stop and that the checker will never get back to idle, and the CheckerPool won't be aware of it, it just knows that it can still create `notCreatedCheckers` more checkers for new requests.
Which means, after <number of cores> requests that create failing checkers, the checkerpool won't create any new checkers and that's how even a request for a simple file fails afterwards, because of all these blocking checkers.

Now that I understand this, I think I nailed down a way to make things even more consistent
- When a checker's cancellation token is triggered, it should just kill the process, period. Maybe give it 2 seconds to recover resources, using Process.Close like it is currently.
- A naïve idea would be that, when the checker's cancellation token is triggered (meaning the checker won't be recovered, and the process is closed), it should also somehow increment `CheckerPool.notCreatedCheckers` so that the CheckerPool is no longer limited. However, this can lead to a deadlock, because if `CheckerPool.notCreatedCheckers` was at 0 right before, it means that any request since the last time it was set to 0 is actually waiting for a checker, and they won't ever recover.
- The robust idea is that, if there is a `checkerWaiter`, it will be immediately fed with a fresh Checker and we don't touch `CheckerPool.notCreatedCheckers`. In the contrary, we increase `CheckerPool.notCreatedCheckers`

Reply:
> > You might say that starting Z3 processes is cheap so we don't need to recycle them, but then why do we have a CheckerPool class in the first place?
> 
> I do not have a definitive answer, but I think this is an excellent question to ask now.

Historically, I think the `CheckerPool` was added for two reasons. One reason was to limit the maximum number of checkers in use. The other reason was to reuse checkers, rather than starting a new Z3 process, because starting a new Z3 process took something like 0.1-0.3 seconds, which was noticeable for files with lots of tiny methods.

Even if we spin up new Z3 processes on demand now, it may be worthwhile to start up some Z3 processes (and put them into the `CheckerPool`) before they are needed. This may reduce the latency in getting a new Z3 process.